In [190]:
import import_ipynb

import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px
from math import sqrt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error

In [191]:
from MLR_CO2_Emission_1 import co2_log, eng_log, cyl_log, fuel_log, fuel_mpg

In [192]:
df2 = pd.DataFrame([eng_log, cyl_log, fuel_log, fuel_mpg, co2_log])

In [193]:
df3 = df2.transpose()
df3

Engine_size  Cylinders  Fuel_cons_comb  Fuel Consumption Comb (mpg)  \
0        0.693147   1.386294        2.140066                     3.496508   
1        0.875469   1.386294        2.261763                     3.367296   
2        0.405465   1.386294        1.774952                     3.871201   
3        1.252763   1.791759        2.406945                     3.218876   
4        1.252763   1.791759        2.360854                     3.295837   
...           ...        ...             ...                          ...   
7375     0.693147   1.386294        2.261763                     3.367296   
7379     0.693147   1.386294        2.197225                     3.433987   
7380     0.693147   1.386294        2.240710                     3.401197   
7382     0.693147   1.386294        2.332144                     3.295837   
7384     0.693147   1.386294        2.370244                     3.258097   

      CO2_emitted  
0        5.278115  
1        5.398163  
2        4.912655  
3        5.541264  
4        5.497168  
...           ...  
7375     5.407172  
7379     5.347108  
7380     5.389072  
7382     5.480639  
7384     5.513429  

[4360 rows x 5 columns]

In [194]:
scalar = StandardScaler()
scalar

StandardScaler()

In [195]:
X = df3[['Engine_size', 'Cylinders', 'Fuel_cons_comb', "Fuel Consumption Comb (mpg)"]]
X

Engine_size  Cylinders  Fuel_cons_comb  Fuel Consumption Comb (mpg)
0        0.693147   1.386294        2.140066                     3.496508
1        0.875469   1.386294        2.261763                     3.367296
2        0.405465   1.386294        1.774952                     3.871201
3        1.252763   1.791759        2.406945                     3.218876
4        1.252763   1.791759        2.360854                     3.295837
...           ...        ...             ...                          ...
7375     0.693147   1.386294        2.261763                     3.367296
7379     0.693147   1.386294        2.197225                     3.433987
7380     0.693147   1.386294        2.240710                     3.401197
7382     0.693147   1.386294        2.332144                     3.295837
7384     0.693147   1.386294        2.370244                     3.258097

[4360 rows x 4 columns]

In [196]:
Y = df3['CO2_emitted']
Y

0       5.278115
1       5.398163
2       4.912655
3       5.541264
4       5.497168
          ...   
7375    5.407172
7379    5.347108
7380    5.389072
7382    5.480639
7384    5.513429
Name: CO2_emitted, Length: 4360, dtype: float64

In [197]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, train_size = 0.8, random_state=1235)

In [198]:
print(X_train.shape, Y_train.shape)
print(X_test.shape, Y_test.shape)

(3488, 4) (3488,)
(872, 4) (872,)


In [199]:
x_train = scalar.fit_transform(X_train)
x_test = scalar.transform(X_test)

In [200]:
model = LinearRegression()
model

LinearRegression()

In [201]:
model_fit = model.fit(x_train, Y_train)
model_fit

LinearRegression()

In [202]:
pred_train = model_fit.predict(x_train)

In [203]:
print("R2 score of log trained model is : ", round(r2_score(pred_train, Y_train),2))

R2 score of log trained model is :  0.9


In [208]:
pred_test = model_fit.predict(x_test)
pred_test

array([5.47823142, 5.3295118 , 5.49696169, 5.33948606, 5.36322931,
       5.68936573, 5.37163764, 5.16547186, 5.37494117, 5.30352456,
       5.43809144, 5.34130039, 5.5794118 , 5.23418496, 5.92139707,
       5.72955736, 5.40141246, 5.44844688, 5.64892632, 5.56552631,
       5.86994781, 5.35833577, 5.55971623, 5.4950374 , 5.27776793,
       5.69086835, 5.32069796, 5.82509726, 5.31672784, 5.62631582,
       5.63843928, 5.37764392, 5.50628592, 5.59024438, 5.39614606,
       5.32924925, 5.50009939, 5.93560549, 6.03175806, 5.35833577,
       5.76896022, 5.84295166, 5.29268435, 5.2777829 , 5.52301212,
       5.56077775, 5.55971623, 4.99168265, 5.58488243, 5.26278715,
       5.75475245, 5.21744068, 5.26278715, 5.28619122, 5.39614606,
       4.89456908, 5.49061667, 5.65501405, 5.76949858, 5.16287212,
       5.21744068, 5.29817313, 5.58278849, 5.23918063, 5.88735889,
       5.48099494, 5.14182934, 5.28289471, 5.25331731, 5.25089923,
       5.35135666, 5.15672772, 5.15532762, 4.98755446, 5.56077

In [209]:
print("R2 score of log tested model is : ", round(r2_score(pred_test, Y_test),2))
print("MAE of log tested model is : ", round(mean_absolute_error(pred_test, Y_test),2))
print("MSE score of log tested model is : ", round(mean_squared_error(pred_test, Y_test),2))
print("RMSE score of log tested model is : ", round(sqrt(mean_squared_error(pred_test, Y_test)),2))

R2 score of log tested model is :  0.91
MAE of log tested model is :  0.04
MSE score of log tested model is :  0.0
RMSE score of log tested model is :  0.07


# Conculsion : Log transformed positively skewed values, along with standardization gives r2 score of 91%, much improved value of 88% from not transformed values.